asset-transfer-basic/rest-api-typescript/src/errors.ts

```diff
@@ -0,0 +1,218 @@
+ *
+ * The regex needs to match the following error messages:
+ *   "the asset %s already exists"
+ *   "The asset ${id} already exists"
+ *   "Asset %s already exists"
+ */
+const matchAssetAlreadyExistsMessage = (message: string): string | null => {
+  //
+  const assetAlreadyExistsRegex = /([tT]he )?[aA]sset \w* already exists/g;
+  const assetAlreadyExistsMatch = message.match(assetAlreadyExistsRegex);
+  logger.debug(
+    { message: message, result: assetAlreadyExistsMatch },
+    'Checking for asset already exists message'
+  );
+
+  if (assetAlreadyExistsMatch !== null) {
+    return assetAlreadyExistsMatch[0];
+  }
+
+  return null;
+};
+
+/*
+ * Matches asset does not exist error strings from the asset contract
+ *
+ * The regex needs to match the following error messages:
+ *   "the asset %s does not exist"
+ *   "The asset ${id} does not exist"
+ *   "Asset %s does not exist"
+ */
+const matchAssetDoesNotExistMessage = (message: string): string | null => {
+  const assetDoesNotExistRegex = /([tT]he )?[aA]sset \w* does not exist/g;
+  const assetDoesNotExistMatch = message.match(assetDoesNotExistRegex);
+  logger.debug(
+    { message: message, result: assetDoesNotExistMatch },
+    'Checking for asset does not exist message'
+  );
+
+  if (assetDoesNotExistMatch !== null) {
+    return assetDoesNotExistMatch[0];
+  }
+
+  return null;
+};
+
+/*
+ * Matches transaction does not exist error strings from the contract API
+ *
+ * The regex needs to match the following error messages:
+ *   "Failed to get transaction with id %s, error Entry not found in index"
+ *   "Failed to get transaction with id %s, error no such transaction ID [%s] in index"
+ */
+const matchTransactionDoesNotExistMessage = (
+  message: string
+): string | null => {
+  const transactionDoesNotExistRegex =
+    /Failed to get transaction with id [^,]*, error (?:(?:Entry not found)|(?:no such transaction ID \[[^\]]*\])) in index/g;
+  const transactionDoesNotExistMatch = message.match(
+    transactionDoesNotExistRegex
+  );
+  logger.debug(
+    { message: message, result: transactionDoesNotExistMatch },
+    'Checking for transaction does not exist message'
+  );
+
+  if (transactionDoesNotExistMatch !== null) {
+    return transactionDoesNotExistMatch[0];
+  }
+
+  return null;
+};
+
+export const isContractError = (err: unknown): boolean => {
+  if (
+    err instanceof AssetExistsError ||
+    err instanceof AssetNotFoundError ||
+    err instanceof TransactionNotFoundError
+  ) {
+    return true;
+  }
+
+  return false;
+};
+
+/*
+ * Handles errors from evaluating and submitting transactions.
+ *
+ * As with duplicate transaction errors, checking error strings like this is
+ * not ideal. Unfortunately the chaincode samples do not use error codes so
+ * again it's the only option. The error message text is not even the same for
+ * the Go, Java, and Javascript implementations of the chaincode!
+ */
```

Comment:
Let's see if we can resolve this in v2.4.

Reply:
There's a limit to what's possible with current chaincode since you only ever get back a message but I'll open an issue and we can have another chat about errors

Reply:
Finally got round to opening that issue! https://github.com/hyperledger/fabric-samples/issues/543